File: AutoBookkeeping/Services/BudgetManager.swift
//
//  BudgetManager.swift
//  AutoBookkeeping
//
//  预算管理服务
//

import Foundation
import SwiftData

@MainActor
class BudgetManager: ObservableObject {

    static let shared = BudgetManager()

    @Published var budgets: [Budget] = []
    @Published var totalBudget: Budget?

    private let modelContext: ModelContext

    private init() {
        self.modelContext = DataManager.shared.getModelContainer().mainContext
        Task {
            await loadBudgets()
        }
    }

    // MARK: - CRUD Operations

    func loadBudgets() async {
        let descriptor = FetchDescriptor<Budget>(
            predicate: #Predicate { $0.isActive },
            sortBy: [SortDescriptor(\.createdAt, order: .reverse)]
        )

        do {
            budgets = try modelContext.fetch(descriptor)
            totalBudget = budgets.first(where: { $0.isTotalBudget })
        } catch {
            print("❌ 加载预算失败: \(error)")
        }
    }

    func saveBudget(_ budget: Budget) async throws {
        modelContext.insert(budget)
        try modelContext.save()
        await loadBudgets()

        // 发送通知
        NotificationCenter.default.post(name: .budgetDidChange, object: nil)
    }

    func updateBudget(_ budget: Budget) async throws {
        budget.updatedAt = Date()
        try modelContext.save()
        await loadBudgets()

        NotificationCenter.default.post(name: .budgetDidChange, object: nil)
    }

    func deleteBudget(_ budget: Budget) async throws {
        modelContext.delete(budget)
        try modelContext.save()
        await loadBudgets()

        NotificationCenter.default.post(name: .budgetDidChange, object: nil)
    }

    // MARK: - Budget Calculations

    /// 获取当前周期的支出
    func getCurrentSpending(
        category: String? = nil,
        period: Budget.BudgetPeriod,
        startDate: Date
    ) async -> Double {
        let dateRange = getDateRange(for: period, startDate: startDate)

        let descriptor: FetchDescriptor<Transaction>
        if let category = category {
            descriptor = FetchDescriptor<Transaction>(
                predicate: #Predicate {
                    $0.type == "expense" &&
                    $0.categoryName == category &&
                    $0.timestamp >= dateRange.start &&
                    $0.timestamp < dateRange.end
                }
            )
        } else {
            descriptor = FetchDescriptor<Transaction>(
                predicate: #Predicate {
                    $0.type == "expense" &&
                    $0.timestamp >= dateRange.start &&
                    $0.timestamp < dateRange.end
                }
            )
        }

        do {
            let transactions = try modelContext.fetch(descriptor)
            return transactions.reduce(0) { $0 + $1.amount }
        } catch {
            print("❌ 获取支出失败: \(error)")
            return 0
        }
    }

    /// 计算预算使用百分比
    func getBudgetUsage(budget: Budget) async -> Double {
        let spending = await getCurrentSpending(
            category: budget.categoryName,
            period: budget.periodEnum,
            startDate: budget.startDate
        )
        return min(spending / budget.amount, 1.0)
    }

    /// 检查是否超支
    func isOverBudget(budget: Budget) async -> Bool {
        let spending = await getCurrentSpending(
            category: budget.categoryName,
            period: budget.periodEnum,
            startDate: budget.startDate
        )
        return spending > budget.amount
    }

    /// 获取超支金额
    func getOverage(budget: Budget) async -> Double {
        let spending = await getCurrentSpending(
            category: budget.categoryName,
            period: budget.periodEnum,
            startDate: budget.startDate
        )
        return max(spending - budget.amount, 0)
    }

    /// 获取剩余预算
    func getRemaining(budget: Budget) async -> Double {
        let spending = await getCurrentSpending(
            category: budget.categoryName,
            period: budget.periodEnum,
            startDate: budget.startDate
        )
        return max(budget.amount - spending, 0)
    }

    /// 获取预算详情
    func getBudgetDetail(budget: Budget) async -> BudgetDetail {
        let spending = await getCurrentSpending(
            category: budget.categoryName,
            period: budget.periodEnum,
            startDate: budget.startDate
        )

        let usage = spending / budget.amount
        let remaining = max(budget.amount - spending, 0)
        let overage = max(spending - budget.amount, 0)
        let isOver = spending > budget.amount

        return BudgetDetail(
            budget: budget,
            spending: spending,
            usage: usage,
            remaining: remaining,
            overage: overage,
            isOver: isOver
        )
    }

    /// 获取所有预算详情
    func getAllBudgetDetails() async -> [BudgetDetail] {
        var details: [BudgetDetail] = []

        for budget in budgets {
            let detail = await getBudgetDetail(budget: budget)
            details.append(detail)
        }

        return details.sorted { !$0.isOver && $1.isOver }
    }

    /// 检查所有预算警告
    func checkBudgetWarnings() async -> [BudgetWarning] {
        var warnings: [BudgetWarning] = []

        for budget in budgets {
            let detail = await getBudgetDetail(budget: budget)

            // 超支警告
            if detail.isOver {
                warnings.append(BudgetWarning(
                    budget: budget,
                    type: .exceeded,
                    message: "\(budget.displayName)已超支 ¥\(detail.overage, specifier: "%.2f")"
                ))
            }
            // 接近预算警告（90%）
            else if detail.usage >= 0.9 {
                warnings.append(BudgetWarning(
                    budget: budget,
                    type: .approaching,
                    message: "\(budget.displayName)已使用 \(Int(detail.usage * 100))%"
                ))
            }
        }

        return warnings
    }

    // MARK: - Helper Methods

    private func getDateRange(for period: Budget.BudgetPeriod, startDate: Date) -> (start: Date, end: Date) {
        let calendar = Calendar.current
        let now = Date()

        switch period {
        case .daily:
            let start = calendar.startOfDay(for: now)
            let end = calendar.date(byAdding: .day, value: 1, to: start)!
            return (start, end)

        case .weekly:
            let start = calendar.dateComponents([.calendar, .yearForWeekOfYear, .weekOfYear], from: now).date!
            let end = calendar.date(byAdding: .weekOfYear, value: 1, to: start)!
            return (start, end)

        case .monthly:
            let components = calendar.dateComponents([.year, .month], from: now)
            let start = calendar.date(from: components)!
            let end = calendar.date(byAdding: .month, value: 1, to: start)!
            return (start, end)

        case .yearly:
            let components = calendar.dateComponents([.year], from: now)
            let start = calendar.date(from: components)!
            let end = calendar.date(byAdding: .year, value: 1, to: start)!
            return (start, end)
        }
    }
}

// MARK: - Supporting Types

struct BudgetDetail {
    let budget: Budget
    let spending: Double
    let usage: Double
    let remaining: Double
    let overage: Double
    let isOver: Bool

    var progressColor: String {
        switch usage {
        case 0..<0.7: return "green"
        case 0.7..<0.9: return "orange"
        default: return "red"
        }
    }
}

struct BudgetWarning {
    let budget: Budget
    let type: WarningType
    let message: String

    enum WarningType {
        case exceeded
        case approaching
    }
}

// MARK: - Notification Names

extension Notification.Name {
    static let budgetDidChange = Notification.Name("budgetDidChange")
}

File: src/services/BudgetManager.ts
/**
 * 预算管理服务
 * 对应 Swift 版本的 BudgetManager.swift
 */

import { EventEmitter } from 'events';
import { Database, Q } from '@nozbe/watermelondb';
import Budget, { BudgetPeriod } from '../models/Budget';
import Transaction from '../models/Transaction';

export interface BudgetDetail {
  budget: Budget;
  spending: number;
  usage: number;
  remaining: number;
  overage: number;
  isOver: boolean;
  progressColor: 'green' | 'orange' | 'red';
}

export interface BudgetWarning {
  budget: Budget;
  type: 'exceeded' | 'approaching';
  message: string;
}

/**
 * 预算管理器 - Singleton
 */
class BudgetManager extends EventEmitter {
  private static instance: BudgetManager | null = null;
  private database!: Database;
  private _budgets: Budget[] = [];
  private _totalBudget: Budget | null = null;

  private constructor() {
    super();
  }

  static getInstance(): BudgetManager {
    if (!BudgetManager.instance) {
      BudgetManager.instance = new BudgetManager();
    }
    return BudgetManager.instance;
  }

  initialize(database: Database) {
    this.database = database;
    this.loadBudgets();
  }

  // MARK: - Getters

  get budgets(): Budget[] {
    return this._budgets;
  }

  get totalBudget(): Budget | null {
    return this._totalBudget;
  }

  // MARK: - CRUD Operations

  async loadBudgets(): Promise<void> {
    try {
      const budgetCollection = this.database.get<Budget>('budgets');
      const budgets = await budgetCollection
        .query(
          Q.where('is_active', true),
          Q.sortBy('created_at', Q.desc)
        )
        .fetch();

      this._budgets = budgets;
      this._totalBudget = budgets.find(b => b.isTotalBudget) || null;

      this.emit('budgetsChanged');
    } catch (error) {
      console.error('❌ 加载预算失败:', error);
    }
  }

  async saveBudget(data: {
    amount: number;
    categoryName?: string;
    period: BudgetPeriod;
    startDate: Date;
    alertThreshold?: number;
  }): Promise<Budget> {
    const budget = await this.database.write(async () => {
      const budgetCollection = this.database.get<Budget>('budgets');
      return await budgetCollection.create(b => {
        b.amount = data.amount;
        b.categoryName = data.categoryName;
        b.period = data.period;
        b.startDate = data.startDate;
        b.alertThreshold = data.alertThreshold || 0.9;
        b.isActive = true;
      });
    });

    await this.loadBudgets();
    this.emit('budgetDidChange');
    return budget;
  }

  async updateBudget(
    budget: Budget,
    updates: {
      amount?: number;
      categoryName?: string | null;
      period?: BudgetPeriod;
      startDate?: Date;
      alertThreshold?: number;
      isActive?: boolean;
    }
  ): Promise<void> {
    await this.database.write(async () => {
      await budget.update(b => {
        if (updates.amount !== undefined) b.amount = updates.amount;
        if (updates.categoryName !== undefined) b.categoryName = updates.categoryName;
        if (updates.period !== undefined) b.period = updates.period;
        if (updates.startDate !== undefined) b.startDate = updates.startDate;
        if (updates.alertThreshold !== undefined) b.alertThreshold = updates.alertThreshold;
        if (updates.isActive !== undefined) b.isActive = updates.isActive;
        b.updatedAt = new Date();
      });
    });

    await this.loadBudgets();
    this.emit('budgetDidChange');
  }

  async deleteBudget(budget: Budget): Promise<void> {
    await this.database.write(async () => {
      await budget.destroyPermanently();
    });

    await this.loadBudgets();
    this.emit('budgetDidChange');
  }

  // MARK: - Budget Calculations

  /**
   * 获取当前周期的支出
   */
  async getCurrentSpending(
    category: string | null | undefined,
    period: BudgetPeriod,
    startDate: Date
  ): Promise<number> {
    const dateRange = this.getDateRange(period, startDate);

    try {
      const transactionCollection = this.database.get<Transaction>('transactions');

      const queries = [
        Q.where('type', 'expense'),
        Q.where('timestamp', Q.gte(dateRange.start.getTime())),
        Q.where('timestamp', Q.lt(dateRange.end.getTime())),
        Q.where('is_deleted', false),
      ];

      if (category) {
        queries.push(Q.where('category_name', category));
      }

      const transactions = await transactionCollection.query(...queries).fetch();

      return transactions.reduce((sum, t) => sum + t.amount, 0);
    } catch (error) {
      console.error('❌ 获取支出失败:', error);
      return 0;
    }
  }

  /**
   * 计算预算使用百分比
   */
  async getBudgetUsage(budget: Budget): Promise<number> {
    const spending = await this.getCurrentSpending(
      budget.categoryName,
      budget.period as BudgetPeriod,
      budget.startDate
    );
    return Math.min(spending / budget.amount, 1.0);
  }

  /**
   * 检查是否超支
   */
  async isOverBudget(budget: Budget): Promise<boolean> {
    const spending = await this.getCurrentSpending(
      budget.categoryName,
      budget.period as BudgetPeriod,
      budget.startDate
    );
    return spending > budget.amount;
  }

  /**
   * 获取超支金额
   */
  async getOverage(budget: Budget): Promise<number> {
    const spending = await this.getCurrentSpending(
      budget.categoryName,
      budget.period as BudgetPeriod,
      budget.startDate
    );
    return Math.max(spending - budget.amount, 0);
  }

  /**
   * 获取剩余预算
   */
  async getRemaining(budget: Budget): Promise<number> {
    const spending = await this.getCurrentSpending(
      budget.categoryName,
      budget.period as BudgetPeriod,
      budget.startDate
    );
    return Math.max(budget.amount - spending, 0);
  }

  /**
   * 获取预算详情
   */
  async getBudgetDetail(budget: Budget): Promise<BudgetDetail> {
    const spending = await this.getCurrentSpending(
      budget.categoryName,
      budget.period as BudgetPeriod,
      budget.startDate
    );

    const usage = spending / budget.amount;
    const remaining = Math.max(budget.amount - spending, 0);
    const overage = Math.max(spending - budget.amount, 0);
    const isOver = spending > budget.amount;

    let progressColor: 'green' | 'orange' | 'red';
    if (usage < 0.7) {
      progressColor = 'green';
    } else if (usage < 0.9) {
      progressColor = 'orange';
    } else {
      progressColor = 'red';
    }

    return {
      budget,
      spending,
      usage,
      remaining,
      overage,
      isOver,
      progressColor,
    };
  }

  /**
   * 获取所有预算详情
   */
  async getAllBudgetDetails(): Promise<BudgetDetail[]> {
    const details: BudgetDetail[] = [];

    for (const budget of this._budgets) {
      const detail = await this.getBudgetDetail(budget);
      details.push(detail);
    }

    // 排序：未超支的在前，超支的在后
    return details.sort((a, b) => {
      if (!a.isOver && b.isOver) return -1;
      if (a.isOver && !b.isOver) return 1;
      return 0;
    });
  }

  /**
   * 检查所有预算警告
   */
  async checkBudgetWarnings(): Promise<BudgetWarning[]> {
    const warnings: BudgetWarning[] = [];

    for (const budget of this._budgets) {
      const detail = await this.getBudgetDetail(budget);

      // 超支警告
      if (detail.isOver) {
        warnings.push({
          budget,
          type: 'exceeded',
          message: `${budget.displayName}已超支 ¥${detail.overage.toFixed(2)}`,
        });
      }
      // 接近预算警告（90%）
      else if (detail.usage >= 0.9) {
        warnings.push({
          budget,
          type: 'approaching',
          message: `${budget.displayName}已使用 ${Math.round(detail.usage * 100)}%`,
        });
      }
    }

    return warnings;
  }

  // MARK: - Helper Methods

  private getDateRange(period: BudgetPeriod, startDate: Date): { start: Date; end: Date } {
    const now = new Date();

    switch (period) {
      case 'daily': {
        const start = new Date(now);
        start.setHours(0, 0, 0, 0);
        const end = new Date(start);
        end.setDate(end.getDate() + 1);
        return { start, end };
      }

      case 'weekly': {
        const start = new Date(now);
        const day = start.getDay();
        const diff = start.getDate() - day + (day === 0 ? -6 : 1); // 调整到周一
        start.setDate(diff);
        start.setHours(0, 0, 0, 0);
        const end = new Date(start);
        end.setDate(end.getDate() + 7);
        return { start, end };
      }

      case 'monthly': {
        const start = new Date(now.getFullYear(), now.getMonth(), 1);
        const end = new Date(now.getFullYear(), now.getMonth() + 1, 1);
        return { start, end };
      }

      case 'yearly': {
        const start = new Date(now.getFullYear(), 0, 1);
        const end = new Date(now.getFullYear() + 1, 0, 1);
        return { start, end };
      }

      default:
        throw new Error(`Unknown period: ${period}`);
    }
  }
}

export default BudgetManager;
